packages/event-cache/src/backends/PostgreSQLBackend.js

```diff
@@ -0,0 +1,205 @@
+     * wrote the SQL and did the data translation myself?
+     *
+     * Do rants count as useful and productive documentation?
+     */
+    this._models = require('../models')
+    this._Sequelize = this._models.Sequelize
+    this._sequelize = this._models.sequelize
+    this._eventTableColumns = Object.keys(this._models.Event.tableAttributes)
+
+    this._loadLatestBlock()
+  }
+
+  /**
+   * Get and set the latest known block from the DB
+   *
+   * @returns {number} The latest block number known by the cache
+   */
+  async _loadLatestBlock() {
+    const result = await this._models.Event.findOne({
+      attributes: [
+        [
+          this._sequelize.fn('MAX', this._sequelize.col('block_number')),
+          'max_block'
+        ]
+      ]
+    })
+    if (result) {
+      const maxBlock = result.dataValues.max_block
+      this.setLatestBlock(maxBlock)
+      debug(`Cache is current up to block #${maxBlock}`)
+      return maxBlock
+    }
+    return 0
+  }
+
+  /**
+   * Convert an arg match object to something that can be used for sequelize
+   * queries as a `where` argument
+   *
+   * @param obj {object} An argMatchObject
+   * @returns {object} to use as a WHERE clause with sequelize
+   */
+  _argMatchToWhere(obj) {
+    const Op = this._Sequelize.Op
+    const newObj = eventToDB(obj)
+    Object.keys(newObj).map(key => {
+      if (!this._eventTableColumns.includes(key)) {
+        newObj['return_values.' + key] = {
+          [Op.eq]: newObj[key]
+        }
+        delete newObj[key]
+      }
+    })
+    return newObj
+  }
+
+  /**
+   * Dumps an array of event objects
+   *
+   * @returns {Array} of events
+   */
+  async serialize() {
+    const res = await this._models.Event.findAll({
+      order: [['block_number'], ['transaction_index'], ['log_index']]
+    })
+    if (res.length < 1) {
+      return []
+    }
+
+    return res.map(row => DBToEvent(row.dataValues))
+  }
+
+  /**
+   * Loads the serialized data from IPFS
+   *
+   * @param ipfsData {Array} An array of events to load
+   */
+  async loadSerialized(ipfsData) {
+    throw new Error('TODO', ipfsData)
+  }
+
+  /**
+   * Fetch events from the store matching objects
+   *
+   * @param argMatchObject {object} A JS object representing the event
+   * @returns {Array} An array of event objects
+   */
+  async get(argMatchObject) {
+    const where = this._argMatchToWhere(argMatchObject)
+    const results = await this._models.Event.findAll({ where })
+    return results.map(row => {
+      return DBToEvent(row.dataValues)
+    })
+  }
+
+  /**
+   * Stores a single event
+   *
+   * For more info on the eventObject, see: https://web3js.readthedocs.io/en/1.0/web3-eth-contract.html#contract-events-return
+   *
+   * @param eventObject {object} A JS object representing the event
+   */
+  async addEvent(eventObject) {
+    const dbObject = eventToDB(eventObject)
+    await this._models.Event.upsert(dbObject)
+    this.setLatestBlock(eventObject.blockNumber)
```

Comment:
This is a general question, not specific to the Postgres implementation.
Is it always the case that events will get added in blockNumber order ?
I wonder for safety purpose if we should check that eventObject.blockNumber > latestBlockNumber before calling setLatestBlock. Or alternatively have that check within the setLatestBlock method,

Reply:
I think for at least the postgres backend, that should be pulled from storage.  That was something I wanted to do but forgot about.

That's significantly less performant for the other backends.  Once the cache is "caught up", that shouldn't be an issue, but I agree in at least adding a check to not allow the block number to be reduced.  (e.g. `if (blockNumber < this.latestBlock) return`)

Thanks, I'll have a fix up this morning